File: AtCoder/abc/abc257/c.cpp
#include <bits/stdc++.h>
#define rep(i, n) for(int i = 0; i < (n); ++i)
#define rep1(i, n) for(int i = 1; i < (n); ++i)
using namespace std;
using ll = long long;

int main(){
    cin.tie(nullptr);
    ios::sync_with_stdio(false);

    int n;
    string s;
    cin >> n >> s;
    vector<int> w(n);
    // vector<pair<int, int>> ws(n);
    map<int, pair<int, int>> count;
    rep(i, n){
        cin >> w[i];
        // ws[i] = {w[i], s[i] - '0'};
        if(s[i] == 0) count[w[i]].first++;
        else count[w[i]].second++;
    }

    int res = 0;
    for(auto c: s){
        if(c == '1') ++res;
    }

    int now = res;
    int ptr = 0;
    auto prev = count.begin();
    for(auto k = count.begin(); k != count.end(); ++k){
        if(ptr == 0){
            prev = k;
            ++ptr;
            continue;
        }
        cout << now << " " << prev->second.first << " " << prev->second.second << endl;
        now += prev->second.first;
        now -= prev->second.second;

        cout << now << endl;
        cout << endl;

        res = max(res, now);        

        prev = k;

        ++ptr;
    }

    // sort(ws.begin(), ws.end());

    // int res = 0;
    // int x = ws.front().first;
    // rep(i, n){
    //     if(ws[i].first >= x && ws[i].second == 1) ++res;
    //     if(ws[i].first < x && ws[i].second == 0) ++res;
    // }

    // int now = res;
    // int i = 1;
    // while(i <= n){
    //     while(ws[i].first == ws[i-1].first){
    //         if(ws[i-1].second == 0) ++now;
    //         else --now;
    //         ++i;
    //     }
    //     res = max(res, now);
    //     ++i;
    // }

    cout << res << endl;

    return 0;
}

File: AtCoder/abc/abc257/c.py
def main():
    n = int(input())
    s = input()
    a = list(map(int, input().split()))

    count = {}
    for i in range(n):
        if a[i] in count:
            if s[i]=='0': count[a[i]][0] += 1
            else: count[a[i]][1] += 1
        else:
            if s[i]=='0': count[a[i]] = [1, 0]
            else: count[a[i]] = [0, 1]

    res = 0
    for c in s:
        if c == '1': res += 1

    ptr = 0
    prev = min(count.keys())
    now = res
    for k in sorted(count.keys()):
        if ptr == 0:
            ptr += 1
            continue

        now += count[prev][0]
        now -= count[prev][1]

        res = max(res, now)
        
        prev = k

    now = 0
    for c in s:
        if c == '0': now += 1
    
    res = max(res, now)

    print(res)        
        

if __name__ == '__main__':
    main()